compiler/AST/foralls.cpp

```diff
@@ -230,3 +235,495 @@ bool astUnderFI(const Expr* ast, ForallIntents* fi) {
+  return NULL;
+}
+
+static void removeOrigIterCall(SymExpr* origSE)
+{
+  INT_ASSERT(!origSE->inTree());
+
+  Symbol* origSym = origSE->symbol();
+  INT_ASSERT(origSym->hasFlag(FLAG_TEMP));
+
+  SymExpr* otherSE = origSym->firstSymExpr();
+  // Only one SE for this symbol
+  INT_ASSERT(origSym->lastSymExpr() == otherSE);
+  
+  if (CallExpr* def = toCallExpr(otherSE->parentExpr))
+    if (def->isPrimitive(PRIM_MOVE))
+      if (otherSE == def->get(1)) {
+        // Remove origSym - except caller will remove origSE.
+        def->remove();
+        origSym->defPoint->remove();
+        return;
+      }
+
+  // The above did not succeed.
+  INT_ASSERT(false);
+}
+
+// Replaces 'origSE' in the tree with the result.
+static CallExpr* buildForallParIterCall(ForallStmt* pfs, SymExpr* origSE)
+{
+  CallExpr* iterCall = NULL;
+
+  if (origSE->symbol()->type->symbol->hasFlag(FLAG_ITERATOR_RECORD)) {
+    // Our iterable expression is an iterator call.
+
+    CallExpr* origIterCall = toCallExpr(findDefOfTemp(origSE));
+    // What to do if we do not find it?
+    INT_ASSERT(origIterCall);
+
+    FnSymbol* origTarget = origIterCall->resolvedFunction();
+    INT_ASSERT(origTarget);
+
+    const char* targetName = origTarget->name;
+    if (!strncmp(targetName, "_parloopexpr", 12))
+      // a forall loop over a (possibly zippered) forall expression, ex.:
+      //  test/reductions/deitz/test_maxloc_reduce_wmikanik_bug2.chpl
+      targetName = astr("_iterator_for_loopexpr", targetName + 12);
+
+    iterCall = origIterCall->copy();
+    iterCall->baseExpr = new UnresolvedSymExpr(targetName);
+
+  } else {
+    // Not an iterator call, so add a call to these().
+    iterCall = new CallExpr("these", gMethodToken, origSE->copy());
+  }
+
+  origSE->replace(iterCall);
+
+  return iterCall;
+}
+
+static bool findStandaloneOrLeader(ForallStmt* pfs,
+                                   CallExpr* iterCall)
+{
+  bool gotParallel = false;
+  bool gotSA = true;
+
+  // We are starting with a serial-iterator call.
+  // Transform it to a standalone/leader call.
+  NamedExpr* tag = new NamedExpr("tag", new SymExpr(gStandaloneTag));
+  iterCall->insertAtTail(tag);
+
+  // try standalone
+  if (!pfs->zippered()) {
+    gotParallel = tryResolveCall(iterCall);
+  }
+
+  // try leader
+  if (!gotParallel) {
+    gotSA = false;
+    tag->actual->replace(new SymExpr(gLeaderTag));
+    gotParallel = tryResolveCall(iterCall);
+  }
+
+  if (!gotParallel) {
+    // Cannot USR_FATAL_CONT in general: e.g. if these() is not found,
+    // we do not know the type of the index variable.
+    // Without which we cannot typecheck the loop body.
+    if (iterCall->isNamed("these") && isTypeExpr(iterCall->get(2)))
+      USR_FATAL(iterCall, "unable to iterate over type '%s'", toString(iterCall->get(2)->getValType()));
+    else
+      USR_FATAL(iterCall, "A%s leader iterator is not found for the iterable expression in this forall loop", pfs->zippered() ? "" : " standalone or");
+  }
```

Comment:
What happens if a leader iterator is defined but a follower isn't ? Where does the error for that come abaout? In some ways I'd expect to see this code resolving both the leader *and* the follower for the leader/follower case.